### PR TITLE
[Bugfix] Bind api server port before starting engine

### LINKED ISSUE #8204: [Bug]: [Errno 98] error while attempting to bind on address ('0.0.0.0', 8000): address already in use
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
Your output of `python collect_env.py` here
```

</details>

### 🐛 Describe the bug

This is a bug we encounter a lot in our ci, e.g. https://buildkite.com/vllm/ci-aws/builds/8098#0191bf43-446d-411d-80c7-3ba10bc392e8/192-1557

I have been tracking this for months, and try to add more logging information to help debugging.

from the logging information:

> [2024-09-05T00:38:34Z] INFO:     Started server process [60858]
> --
>   | [2024-09-05T00:38:34Z] INFO:     Waiting for application startup.
>   | [2024-09-05T00:38:34Z] INFO:     Application startup complete.
>   | [2024-09-05T00:38:34Z] ERROR:    [Errno 98] error while attempting to bind on address ('0.0.0.0', 44319): [errno 98] address already in use
>   | [2024-09-05T00:38:34Z] INFO:     Waiting for application shutdown.
>   | [2024-09-05T00:38:34Z] INFO:     Application shutdown complete.
>   | [2024-09-05T00:38:34Z] DEBUG 09-04 17:38:34 launcher.py:64] port 44319 is used by process psutil.Process(pid=60914, name='pt_main_thread', status='sleeping', started='17:37:05') launched with command:
>   | [2024-09-05T00:38:34Z] DEBUG 09-04 17:38:34 launcher.py:64] /usr/bin/python3 -c from multiprocessing.spawn import spawn_main; spawn_main(tracker_fd=16, pipe_handle=18) --multiprocessing-fork
> 
> 

we can see that the server process is pid 60858 , and the port 44319 is used by process 60914. scrolling up a little bit, we can find:

> [2024-09-05T00:37:05Z] INFO 09-04 17:37:05 api_server.py:160] Multiprocessing frontend to use ipc:///tmp/b6851f4d-4d78-46b8-baba-ae179b0088c2 for RPC Path.
> --
>   | [2024-09-05T00:37:05Z] INFO 09-04 17:37:05 api_server.py:176] Started engine process with PID 60914
> 

it becomes clear that this is the engine process.

I think the problem here, is that we only bind the port after the engine is ready. During engine setup, it might use some ports for ray, or for distributed communication.

there are two possible solutions:
1. the api server immediately binds to the port after start, and returns unready status when client queries the `/healthy` endpoint
2. the api server binds the port immediately (via `socket.socket(socket.AF_INET, socket.SOCK_STREAM).bind(("", uvicorn_kwargs["port"]))`), and after engine is up, it releases the port, and bind again to serve requests

I think 1 might be better. 2 would suffer from the fact that client will get 404 not found before the engine is up, because this is just a raw socket without any response.

cc @robertgshaw2-neuralmagic @njhill @joerunde 

also cc @richardliaw @rkooo567 how to turn on verbose ray logging, so that we can verify if the port is indeed used by ray.

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by kevin314**
~Potential alternate solution for #8204--I saw that asyncio will use a default port of 0 in the context of create_server() if no port is given in the config. Binding to port 0 with socket.socket.bind() apparently means the OS will pick an available port for you, which I think functions similarly to get_open_port()~

~But instead of choosing the port first and binding it later after the engine starts, choosing as the port is bound may resolve potential conflicts with Ray picking the same port?~

Bind the chosen api server port to a temp socket until engine is started to avoid port conflict.

FIX #8204

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by youkaichao** [comment]
but then how does the caller know which port to connect to the api server?

**Comment by youkaichao** [comment]
one way is to read and parse the output, like `Uvicorn running on http://0.0.0.0:40243` . but it is not elegant.

**Comment by youkaichao** [comment]
I'm curious why I don't see any users complaining about this. And I think the answer is, users use port `8000` by default. It does not overlap with dynamically allocated ports.

for normal os, when we ask for a random port, it is not really a random port. it is chosen from some range, e.g.

```shell
$ cat /proc/sys/net/ipv4/ip_local_port_range
32768   60999
```

And given the fact that we never launch multiple api servers, **we might solve this issue by directly assign one specific port for it, say `8001`** 😄

**Comment by kevin314** [comment]
@youkaichao Oh I see. So even though this issue is for automated testing, sometimes users would still want to connect to this test server manually to debug or something? And they just happened to rarely run into a situation where port 8000 was taken, and have to check 8001 or 8002.

**Comment by youkaichao** [comment]
> So even though this issue is for automated testing, sometimes users would still want to connect to this test server manually to debug or something?

for testing, we cannot just start the server and then kill it. we need to use a client to connect to the server and let it run something.

**Comment by kevin314** [comment]
@youkaichao Ok, I see where my confusion was now with the client needing to know the server's port. I think I'll actually try recreating the issues seen in the CI locally, and then try out some of these other solutions to verify they work. Thanks!

**Comment by kevin314** [comment]
Also, would you rather I create a new PR, or is it ok if I just modify this same draft?

**Comment by SolitaryThinker** [comment]
Just modify this one and change the title

**Comment by youkaichao** [comment]
see https://github.com/vllm-project/vllm/blob/3724d5f6b59d9859e5b47c047535bb8edc124eab/tests/utils.py#L144-L162

the `self.port` is used to creat a client.

**Review by youkaichao — APPROVED** [review]
LGTM, thanks for the fix! let's see if this can pass all the tests.

**Comment by youkaichao** [comment]
the result looks promising! can you resolve the conflict so that we can merge?

**Comment by youkaichao** [comment]
merge as error is unrelated

File: vllm/entrypoints/openai/api_server.py [hunk 1]
```diff
@@ -525,6 +526,9 @@ async def run_server(args, **uvicorn_kwargs) -> None:
     logger.info("vLLM API server version %s", VLLM_VERSION)
     logger.info("args: %s", args)
 
+    temp_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    temp_socket.bind(("", args.port))
```
Comment (piamo) [inline vllm/entrypoints/openai/api_server.py:530]: this may disable listening on ipv6 address

**Pull request merged**